### PR TITLE
Fix illegal blastmining drops

### PR DESCRIPTION
**Pull request by RasmusKD**
Fixed so now spawners actually dont drop from blastmining, mining wont spawn illegal items anymore, before items like suspisous sand, infested_stone and similar illegal items were requiable with blastmining.

Removed double drops from non-mining related blocks Added stone types back into the drops w/o double drops 

Added Javadoc

### CODE REVIEW

File: src/main/java/com/gmail/nossr50/util/ItemUtils.java [hunk 1]
```diff
@@ -147,18 +149,29 @@ public static boolean hasItemIncludingOffHand(Player player, Material material)
...
      */
     public static void removeItemIncludingOffHand(@NotNull Player player, @NotNull Material material, int amount) {
-        // Checks main inventory / item bar
-        if (player.getInventory().contains(material)) {
-            player.getInventory().removeItem(new ItemStack(material, amount));
-            return;
+        // Remove from main inventory
+        for (ItemStack itemStack : player.getInventory().getContents()) {
+            if (itemStack != null && itemStack.getType() == material) {
+                int stackAmount = itemStack.getAmount();
```
Comment (nossr50) [inline src/main/java/com/gmail/nossr50/util/ItemUtils.java:155]: nit: could be final

File: src/main/java/com/gmail/nossr50/util/ItemUtils.java [hunk 2]
```diff
@@ -147,18 +149,29 @@ public static boolean hasItemIncludingOffHand(Player player, Material material)
      * @param amount   Amount of the material to remove
      */
     public static void removeItemIncludingOffHand(@NotNull Player player, @NotNull Material material, int amount) {
-        // Checks main inventory / item bar
-        if (player.getInventory().contains(material)) {
-            player.getInventory().removeItem(new ItemStack(material, amount));
```
Comment (nossr50) [inline src/main/java/com/gmail/nossr50/util/ItemUtils.java:152]: Could you please explain why you had to replace lines 150-153? It does not seem obvious why the code you've replaced it with would do anything differently.

File: src/main/java/com/gmail/nossr50/util/ItemUtils.java [hunk 3]
```diff
@@ -147,18 +149,29 @@ public static boolean hasItemIncludingOffHand(Player player, Material material)
...
 
-        // Check off-hand
-        final ItemStack offHandItem = player.getInventory().getItemInOffHand();
-        if (offHandItem.getType() == material) {
-            int newAmount = offHandItem.getAmount() - amount;
-            if (newAmount > 0) {
-                offHandItem.setAmount(newAmount);
+        // Remove from off-hand
+        ItemStack offHandItem = player.getInventory().getItemInOffHand();
+        if (offHandItem != null && offHandItem.getType() == material) {
```
Comment (nossr50) [inline src/main/java/com/gmail/nossr50/util/ItemUtils.java:171]: Was this to avoid a NPE or something else?

File: src/util/java/mcMMO/PotionConfigGenerator.java [hunk 4]
```diff
@@ -104,6 +104,8 @@ private static String getMCName(PotionType type) {
                 case WEAKNESS :
                 case TURTLE_MASTER:
                 case SLOW_FALLING:
+                case INFESTED:
```
Comment (nossr50) [inline src/util/java/mcMMO/PotionConfigGenerator.java:107]: This won't be compatible with older versions, you'll have to change the switch case to operate on string value

File: src/util/java/mcMMO/PotionConfigGenerator.java
Comment (nossr50) [inline src/util/java/mcMMO/PotionConfigGenerator.java:1]: Side note: PotionConfigGenerator disgusts me 😆 , I appreciate that you went through the trouble to update it.

File: src/main/java/com/gmail/nossr50/skills/mining/MiningManager.java [hunk 5]
```diff
@@ -31,55 +31,78 @@
...
+     *
+     * @return true if the player can use Demolitions Expertise, false otherwise
+     */
     public boolean canUseDemolitionsExpertise() {
-        if (!RankUtils.hasUnlockedSubskill(getPlayer(), SubSkillType.MINING_DEMOLITIONS_EXPERTISE))
-            return false;
-
-        return getSkillLevel() >= BlastMining.getDemolitionExpertUnlockLevel() && Permissions.demolitionsExpertise(getPlayer());
+        return RankUtils.hasUnlockedSubskill(getPlayer(), SubSkillType.MINING_DEMOLITIONS_EXPERTISE)
+                && getSkillLevel() >= BlastMining.getDemolitionExpertUnlockLevel()
```
Comment (nossr50) [inline src/main/java/com/gmail/nossr50/skills/mining/MiningManager.java:50]: this skill level check is unnecessary

File: src/main/java/com/gmail/nossr50/skills/mining/MiningManager.java [hunk 6]
```diff
@@ -31,55 +31,78 @@
...
+     *
+     * @return true if the player can use Bigger Bombs, false otherwise
+     */
     public boolean canUseBiggerBombs() {
-        if (!RankUtils.hasUnlockedSubskill(getPlayer(), SubSkillType.MINING_BIGGER_BOMBS))
-            return false;
-
-        return getSkillLevel() >= BlastMining.getBiggerBombsUnlockLevel() && Permissions.biggerBombs(getPlayer());
+        return RankUtils.hasUnlockedSubskill(getPlayer(), SubSkillType.MINING_BIGGER_BOMBS)
+                && getSkillLevel() >= BlastMining.getBiggerBombsUnlockLevel()
```
Comment (nossr50) [inline src/main/java/com/gmail/nossr50/skills/mining/MiningManager.java:83]: Skill level check is unnecessary

File: src/main/java/com/gmail/nossr50/skills/mining/MiningManager.java [hunk 7]
```diff
@@ -175,66 +210,56 @@ public void remoteDetonation() {
      * @param event The {@link EntityExplodeEvent}
      */
     public void blastMiningDropProcessing(float yield, EntityExplodeEvent event) {
-        if (yield == 0)
-            return;
+        if (yield == 0) return;
```
Comment (nossr50) [inline src/main/java/com/gmail/nossr50/skills/mining/MiningManager.java:213]: not a fan of this change, would like it changed back

File: src/main/java/com/gmail/nossr50/skills/mining/MiningManager.java [hunk 8]
```diff
@@ -175,66 +210,56 @@ public void remoteDetonation() {
...
 
         var increasedYieldFromBonuses = yield + (yield * getOreBonus());
-        // Strip out only stuff that gives mining XP
```
Comment (nossr50) [inline src/main/java/com/gmail/nossr50/skills/mining/MiningManager.java:182]: Deleting comments is an unusual thing to do in a PR, try to keep your diff minimal

File: src/main/java/com/gmail/nossr50/skills/mining/MiningManager.java [hunk 9]
```diff
@@ -31,55 +31,78 @@
...
+     * Determines if the player can detonate TNT remotely.
+     *
+     * @return true if the player can detonate TNT remotely, false otherwise
+     */
     public boolean canDetonate() {
         Player player = getPlayer();
-
         return canUseBlastMining() && player.isSneaking()
-                && (isPickaxe(getPlayer().getInventory().getItemInMainHand()) || player.getInventory().getItemInMainHand().getType() == mcMMO.p.getGeneralConfig().getDetonatorItem())
+                && (isPickaxe(player.getInventory().getItemInMainHand())
```
Comment (nossr50) [inline src/main/java/com/gmail/nossr50/skills/mining/MiningManager.java:62]: I appreciate trying to clean up the mess that is the existing code, but try to keep the diff minimal for review

File: src/main/java/com/gmail/nossr50/skills/mining/MiningManager.java [hunk 10]
```diff
@@ -175,66 +210,56 @@ public void remoteDetonation() {
...
         List<BlockState> ores = new ArrayList<>();
-        List<BlockState> notOres = new ArrayList<>();
+        List<BlockState> nonOres = new ArrayList<>();
+
         for (Block targetBlock : event.blockList()) {
             BlockState blockState = targetBlock.getState();
 
-            if(mcMMO.getUserBlockTracker().isIneligible(targetBlock))
-                continue;
+            if (mcMMO.getUserBlockTracker().isIneligible(targetBlock)) continue;
```
Comment (nossr50) [inline src/main/java/com/gmail/nossr50/skills/mining/MiningManager.java:223]: Please don't put executions on the same line as `if`s, not a fan.

**Review by nossr50 — CHANGES_REQUESTED** [review]
I find your changes to `MiningManager` a bit hard to follow, please minimize the diff and I'll review again.

I would also prefer you to split these commits into their own PRs if possible.

**Comment by nossr50** [comment]
PR title seems a bit odd, if you are going to keep it as 3 commits and one PR (which I would prefer you did not), I would like the title of the PR changed to reflect the changes, please add a description of all changes in the PR as well.

File: src/main/java/com/gmail/nossr50/skills/mining/MiningManager.java (hunk 5 above)
Comment (RasmusKD) [reply src/main/java/com/gmail/nossr50/skills/mining/MiningManager.java:50]: It was solely to streamline the condition so you want me to revert it?
